File: app/src/main/java/com/ptithcm/thuan6420/basecleanarchitecture/ui/login/RegisterFragment.java
package com.ptithcm.thuan6420.basecleanarchitecture.ui.login;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;

import com.ptithcm.thuan6420.basecleanarchitecture.R;
import com.ptithcm.thuan6420.basecleanarchitecture.databinding.FragmentRegisterBinding;

public class RegisterFragment extends Fragment implements View.OnClickListener {

    private FragmentRegisterBinding binding;
    private UserController userController;

    private void initController() {
        userController = new UserController(this);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        binding = FragmentRegisterBinding.inflate(inflater, container, false);
        initController();
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        binding.tvToLogin.setOnClickListener(this);
        binding.btnRegister.setOnClickListener(this);
        binding.layoutRegisterFragment.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        userController.handleOnClick(v);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        binding = null;
    }

    public void getValue() {
        userController.setUserEmail(binding.etEmailRegister.getText().toString().trim());
        userController.setUserPassword(binding.etPasswordRegister.getText().toString().trim());
        userController.setUserFullName(binding.etFullNameRegister.getText().toString().trim());
        userController.setUserPhoneNumber(binding.etPhoneNumberRegister.getText().toString().trim());
    }

    public void showErrorEmail(String errorMessage) {
        binding.etEmailRegisterLayout.setError(errorMessage);
    }

    public void showErrorPassword(String errorMessage) {
        binding.etPasswordRegisterLayout.setError(errorMessage);
    }

    public void showErrorFullName(String errorMessage) {
        binding.etFullNameRegisterLayout.setError(errorMessage);
    }

    public void showErrorPhoneNumber(String errorMessage) {
        binding.etPhoneNumberRegisterLayout.setError(errorMessage);
    }

    public void navigateToLoginFragment() {
        NavHostFragment.findNavController(RegisterFragment.this)
                .navigate(R.id.action_registerFragment_to_loginFragment);
    }

    public void turnOnLoading() {
        binding.progressBarRegister.setVisibility(View.VISIBLE);
    }

    public void turnOffLoading() {
        binding.progressBarRegister.setVisibility(View.GONE);
    }

    public void setEnabledButton(boolean enabled) {
        binding.btnRegister.setEnabled(enabled);
    }
}

File: app/src/main/java/com/ptithcm/thuan6420/basecleanarchitecture/ui/login/RegisterFragment.kt
package com.ptithcm.thuan6420.basecleanarchitecture.ui.login

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.NavHostFragment
import com.ptithcm.thuan6420.basecleanarchitecture.R
import com.ptithcm.thuan6420.basecleanarchitecture.ui.components.UiController
import com.ptithcm.thuan6420.basecleanarchitecture.ui.dialogs.DialogController
import com.ptithcm.thuan6420.basecleanarchitecture.ui.dialogs.IClickOnButtonDialogListener

class RegisterFragment : Fragment(), RegisterListener, IClickOnButtonDialogListener {
    private lateinit var mRegisterView: RegisterView

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        mRegisterView = RegisterViewImp(inflater, container)
        return mRegisterView.getRootView()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mRegisterView.setListener(this)
    }

    fun navigateToLoginFragment() {
        NavHostFragment.findNavController(this@RegisterFragment)
            .navigate(R.id.action_registerFragment_to_loginFragment)
    }

    override fun clickOnButtonDialog(id: Int) {
        if (id == R.id.btnSuccess) {
            navigateToLoginFragment()
        }
    }

    override fun onClickLayout() {
        UiController().closeKeyBoard(this.requireActivity())
    }

    override fun onClickButtonRegister(
        invalidFields: List<InvalidField>?,
        email: String,
        password: String,
        fullName: String,
        phoneNumber: String
    ) {
        var countErrorField: Byte = 0
        for (invalidField in invalidFields!!) {
            if (invalidField.message != null) {
                countErrorField++
            }
        }
        if (countErrorField == UserController().NO_FIELD) {
            val responseUI =
                UserController().isRegisterSuccess(email, password, fullName, phoneNumber)

            if (responseUI.success) {
                DialogController().showSuccessDialog(this.context, this, responseUI.message)
            } else {
                DialogController().showErrorDialog(this.context, this, responseUI.message)
            }
        }
    }

    override fun onClickTvToRegister() {
        navigateToLoginFragment()
    }

    override fun findInvalidFields(
        email: String,
        password: String,
        fullName: String,
        phoneNumber: String
    ): List<InvalidField> =
        UserController().findInvalidRegisterFields(email, password, fullName, phoneNumber)
}